spam/spam_tutorial.py

```diff
@@ -140,234 +145,406 @@
+
+# %%
+import re
+
+
+@labeling_function()
+def regex_check_out(x):
+    return SPAM if re.search(r"check.*out", x.text, flags=re.I) else ABSTAIN
+
+
+# %% [markdown]
+# Again, let's generate our label matrices and see how we do.
+
+# %%
+lfs = [check_out, check, regex_check_out]
+
+applier = PandasLFApplier(lfs=lfs)
+L_train = applier.apply(df=df_train)
+L_dev = applier.apply(df=df_dev)
+
+# %%
+LFAnalysis(L=L_train, lfs=lfs).lf_summary()
+
+# %%
+LFAnalysis(L_dev, lfs).lf_summary(Y_dev)
+
+# %% [markdown]
+# We've split the difference in `train` set coverage, and increased our accuracy on the `dev` set to 100%!
+# This looks promising.
+# Let's verify that we corrected our false positive from before.
+
+# %%
+buckets = error_buckets(L_dev[:, 1], L_dev[:, 2])
+df_dev.iloc[buckets[(ABSTAIN, SPAM)]]
+
+# %% [markdown]
+# To understand the coverage difference between `check` and `regex_check_out`, let's take a look at the training set.
+# Remember: coverage isn't always good.
+# Adding false positives will increase coverage.
+
+# %%
+buckets = error_buckets(L_train[:, 1], L_train[:, 2])
+df_train.iloc[buckets[(ABSTAIN, SPAM)]].sample(10, random_state=1)
+
+# %% [markdown]
+# Most of these are SPAM, but a good number are false positives.
+# **To keep precision high (while not sacrificing much in terms of coverage), we'd choose our regex-based rule.**
+
+# %% [markdown]
+# ### e) Writing an LF that uses a third-party model
+
+# %% [markdown]
+# The LF interface is extremely flexible, and can wrap existing models.
+# A common technique is to use a commodity model trained for other tasks that are related to, but not the same as, the one we care about.
 #
-# The helper method `error_buckets()` groups examples by their predicted label and true label. For example, `buckets[(SPAM, HAM)]` contains the indices of data points that the LF labeled `SPAM` that actually belong to class `HAM`. This may give ideas for where the LF could be made more specific.
+# For example, the [TextBlob](https://textblob.readthedocs.io/en/dev/index.html) tool provides a pretrained sentiment analyzer. Our spam classification task is not the same as sentiment classification, but it turns out that `SPAM` and `HAM` comments have different distributions of sentiment scores.
+# We'll focus on writing LFs for `HAM`, since we identified `SPAM` comments above.
+#
+# **A brief intro to `Preprocessor`s**
+#
+# Just like a `LabelingFunction` is constructed from a black-box Python function that maps a data point to an integer label,
+# a `Preprocessor` is a black-box Python function that maps a data point to a new data point.
+# `LabelingFunction`s can use `Preprocessor`s, which lets us write LFs over transformed or enhanced data points.
+# We use the `@preprocessor(...)` decorator to add extra functionality to preprocessing functions, such as memoization (i.e. input/output caching, so it doesn't re-execute for each LF that uses it).
```

Comment:
Here is sounds like the extra functionality of the decorator is optional and the user can just pass a function to the preprocessors argument if they don't need that functionality.

You can instead say 'preprocessors are created using the preprocessor() decorator' like you did for LFs, and then mention memoization as a bonus.